src/nools/lib.js

```diff
@@ -1,5 +1,34 @@
+function bindAllFunctionsToContext(obj, context) {
+  var keys = Object.keys(obj);
+  for (var i in keys) {
+    var key = keys[i];
+    switch(typeof obj[key]) {
+      case 'object':
+        bindAllFunctionsToContext(obj[key], context);
+        break;
+      case 'function':
+        obj[key] = obj[key].bind(context);
+        break;
+    }
+  }
+}
+
+function deepCopy(obj) {
+  var copy = Object.assign({}, obj);
+  var keys = Object.keys(copy);
+  for (var i in keys) {
+    var key = keys[i];
+    if (typeof copy[key] === 'object') {
+      copy[key] = deepCopy(copy[key]);
+    }
+  }
+  return copy;
+}
+
```

Comment:
This function doesn't handle arrays well, unfortunately: 
```
console.log(deepCopy({ a: [1, 2, 3] })) 
> { a: { '0': 1, '1': 2, '2': 3 } }
```

Reply:
Good catch

Reply:
There are still a couple of edge cases where this function fails:
```
console.log(deepCopy([1, 2, 3]));
> { '0': 1, '1': 2, '2': 3 }
console.log(deepCopy({a: [[1], [2], [3]]}));
> { a: [ { '0': 1 }, { '0': 2 }, { '0': 3 } ] }